src/payment/models/buy-crypto/services/buy-crypto.service.ts

```diff
@@ -176,8 +178,52 @@ export class BuyCryptoService {
     });
   }
 
+  async getHistory(userId: number, buyId: number): Promise<BuyCryptoHistoryDto[]> {
+    return this.buyCryptoRepo
+      .find({
+        where: { buy: { id: buyId, user: { id: userId } } },
+        relations: ['buy', 'buy.user'],
+      })
+      .then((buyCryptos) => buyCryptos.map(this.toHistoryDto));
+  }
+
+  async getCryptoRouteHistory(userId: number, routeId: number): Promise<CryptoRouteHistoryDto[]> {
+    return this.buyCryptoRepo
+      .find({
+        where: { cryptoRoute: { id: routeId, user: { id: userId } } },
+        relations: ['cryptoInput', 'cryptoRoute', 'cryptoRoute.user'],
+      })
+      .then((history) => history.map(this.toCryptoRouteHistoryDto));
+  }
+
   // --- HELPER METHODS --- //
 
+  private toHistoryDto(buyCrypto: BuyCrypto): BuyCryptoHistoryDto {
+    return {
+      inputAmount: buyCrypto.inputAmount,
+      inputAsset: buyCrypto.inputAsset,
+      amlCheck: buyCrypto.amlCheck,
+      outputAmount: buyCrypto.outputAmount,
+      outputAsset: buyCrypto.outputAsset,
+      txId: buyCrypto.txId,
+      isComplete: buyCrypto.isComplete,
+      date: buyCrypto.outputDate,
+    };
+  }
+
+  private toCryptoRouteHistoryDto(buyCrypto: BuyCrypto): CryptoRouteHistoryDto {
+    return {
+      inputAmount: buyCrypto.cryptoInput.amount,
+      inputAsset: buyCrypto.cryptoInput.asset.dexName,
+      outputAmount: buyCrypto.outputAmount,
+      outputAsset: buyCrypto.outputAsset,
+      txId: buyCrypto.txId,
+      date: buyCrypto.outputDate,
+      amlCheck: buyCrypto.amlCheck,
+      isComplete: buyCrypto.isComplete,
+    };
+  }
```

Comment:
I think you can use the `toHistoryDto` method for crypto routes as well, can't you?

Reply:
If yes, you could also remove the cryptoInput relation in getCryptoRouteHistory.

Reply:
No, I can't. Judging based on the test data that is around. I need `cryptoInput` to get the real input in BTC. inside `BuyCrypto` the input amount is set in EUR and therefore weird to show, if the user has sent in BTC and suddenly we show EUR.

But might be misunderstand from my side or test data being incorrect.

Reply:
Oh I see. On the Dev system, BuyCryptos are not automatically filled. We therefore just entered some random information do to the test. That probably was a bad idea 🤔 

But that means that on production you'll have BTC as an input asset.

Reply:
Okay that is something I am not aware. But if it is the case, I am happy to use the same function and remove the crypto one